internal/cluster/cluster.go

```diff
@@ -745,6 +748,7 @@ func (m *Manager) CancelInstallation(ctx context.Context, c *common.Cluster, rea
 		if funk.Contains(installationStates, lastState.srcState) {
 			m.metricAPI.ClusterInstallationFinished(ctx, models.ClusterStatusCancelled, lastState.srcState, c.OpenshiftVersion, *c.ID, c.EmailDomain, c.InstallStartedAt)
 		}
+		go m.UploadEvents(ctx, c)
```

Comment:
There is another good reason to avoid it (although I was the one to suggest it) - since the recent events aren't committed to the DB at this point (the transaction may be still open), they won't be collected.
Seems like a safer option is to push the call to `m.UploadEvents` to a later phase.

Reply:
Thanks both of you for catching this! 

Maybe an alternative option is to call it from here? https://github.com/openshift/assisted-service/blob/445e8646e95c70b74c9131c57945eff1289f73da/internal/bminventory/inventory.go#L3817
After the transaction completes? 


I'll look into what @slaviered suggested too - to make background goroutine and call it from a channel

Reply:
Another option can be to have another thread as Sarah suggested just instead of using channels it can look at the DB for clusters that got into a final state, search events basically, and it can have a small table to track on which clusters it already uploaded. 
Moti, Sarah what do you think?

Reply:
Sounds better, as it decouples the execution of uploading the events from this context.

Reply:
I agree as well. The monitor thread is a good example on how to do it.

Reply:
Thank you all for the feedback, suggestions, and ideas! I've gone ahead and tried to implement the option @filanov suggested. 

Changes:
1. Started an uploader thread in main as @slaviered suggested https://github.com/openshift/assisted-service/pull/4880/files#diff-c444f711e9191b53952edb65bfd8c644419fc7695c62611dc0fb304b4fb197d6R390-R394
2. Modified the `UploadEvents` function that's called in the thread https://github.com/openshift/assisted-service/pull/4880/files#diff-ecbdbfaa3e0752590ccb7b5c0387c308de661ba622120129edf662c02fa1ddadR923 
    - This function checks all cluster statuses and only uploads for clusters that: 
        - haven't already been uploaded and 
        - its status is one of [`Installed`, `error`, `cancelled`] or ([`Adding Hosts`] AND `InstallTimestamp` is not zero)
4. Added a new DB table to track which clusters have been uploaded https://github.com/openshift/assisted-service/pull/4880/files#diff-01bdb7dd494494bebd086fc6d8172ab9a4da151e7fb194a40616bfba462a27e4R25-R28 (naming suggested appreciated) along with the following changes to support this table:
    - Creating an entry in the table during `registerCluster` https://github.com/openshift/assisted-service/pull/4880/files#diff-268adb06c81106332ee620da395932a03fdf3059b29acc0dbcd66d6cf484f8e9R74
    - Deleting the entry during `DeregisterCluster` https://github.com/openshift/assisted-service/pull/4880/files#diff-268adb06c81106332ee620da395932a03fdf3059b29acc0dbcd66d6cf484f8e9R148
    - Updating the entry if the data was uploaded https://github.com/openshift/assisted-service/pull/4880/files#diff-ecbdbfaa3e0752590ccb7b5c0387c308de661ba622120129edf662c02fa1ddadR970-R974

Please let me know what you think. Especially with regards the DB items!